### PR TITLE
tests: Add bt_id_read_public_addr() unit tests

### PR DESCRIPTION
**Pull request by ahmedmoheb-nordic**
Unit test project for bt_id_read_public_addr().
This part of subsys/bluetooth/host/id.c unit testing.

### CODE REVIEW

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+FILE(GLOB SOURCES src/*.c)
+
+project(bt_id_read_public_addr)
+add_compile_definitions(test_unit_name="${PROJECT_NAME}")
+
+# Suppress the format-zero-length error if GNUC is used
+if (CMAKE_C_COMPILER_ID MATCHES "GNU")
+add_compile_options(-Wno-error=format-zero-length -Wno-format-zero-length)
+endif()
```
Comment (alwa-nordic) [inline tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:13]: Why is this needed? I don't think we should have this.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+FILE(GLOB SOURCES src/*.c)
```
Comment (alwa-nordic) [inline tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:5]: Please don't use glob in cmake files. It makes the build system unreliable.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+project(bt_id_read_public_addr)
+add_compile_definitions(test_unit_name="${PROJECT_NAME}")
```
Comment (alwa-nordic) [inline tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:8]: Don't do this. Instead, spell out the function name in the messages where this is used.

File: tests/bluetooth/host/id/bt_id_read_public_addr/src/main.c [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ *
+ *  Expected behaviour:
+ *   - Execution stops and a zero return value is returned which represents failure
+ */
+ZTEST(bt_id_read_public_addr, test_bt_hci_cmd_send_sync_returns_err)
+{
+	int err;
+	bt_addr_le_t addr;
+
+	bt_hci_cmd_send_sync_fake.return_val = !0;
```
Comment (alwa-nordic) [inline tests/bluetooth/host/id/bt_id_read_public_addr/src/main.c:53]:
```suggestion
	bt_hci_cmd_send_sync_fake.return_val = 1;
```

**Review by alwa-nordic — COMMENTED** [review]
Please run clang-format on all new code.

You are missing a test confirming that the address is copied to the output parameter on success.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
See comments above.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt (hunk 1 above)
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:13]: Please have a look at this discussion https://github.com/zephyrproject-rtos/zephyr/pull/51329#discussion_r997767844

This is a workaround to the following warning `warning: zero-length gnu_printf format string [-Wformat-zero-length] ` generating while compiling `BT_DBG("");` for unit testing

File: tests/bluetooth/host/id/bt_id_read_public_addr/src/main.c (hunk 4 above)
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/src/main.c:53]: Ok, but it was intended to be read as `NOT 0`

**Comment by ahmedmoheb-nordic** [comment]
> Please run clang-format on all new code.
> 
> You are missing a test confirming that the address is copied to the output parameter on success.

Thanks for the note for not checking if the address is copied to the output parameter, I fixed it.

Regarding the note for running clang-format on new sources, Is it a generic remark? or you already found non-formatted coded.
I ran it on current source files and there were no changes on the code except it's re-ordering the header files inclusions in a way that no project use. 
I discarded changes related to header files inclusion re-ordering, please let me know if you think that I should commit the result from clang-format in anyway.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt (hunk 3 above)
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:8]: Most of the error messages are the same for different tests except that they differs in the API under testing.
I was looking for a way to unify the printed messages structure.

Ok, I will remove such definition from `CMakeLists.txt` files and use `__func__` which combines the test suite and test case names.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt (hunk 2 above)
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:5]: What do you think about the following for replacing `glob`?

```suggestion
set(SOURCES
  ${CMAKE_CURRENT_SOURCE_DIR}/src/main.c
  ${CMAKE_CURRENT_SOURCE_DIR}/src/test_suite_invalid_inputs.c
)

file(GLOB files src/*.c)
foreach(file ${files})
list (FIND SOURCES ${file} _index)
if(_index EQUAL -1)
  message(WARNING "File ${file} not found in sources list!")
endif()
endforeach()
```
The checking part is just for printing a warning if there are source files in the `src` directory that don't exist in the source files list.

Comment (alwa-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:5]: I am OK with this as an experiment. The worst thing to happen would be a false warning. But we have to use `file(GLOB files CONFIGURE_DEPENDS src/*.c)` to make it a bit more reliable.

I don't like the`${CMAKE_CURRENT_SOURCE_DIR}/` in the sources list. Requiring it changes the semantics of the `SOURCES` variable (compared to the rest of Zephyr). We should be able to use relative paths there. I think we should allow relative paths and use `file(REAL_PATH ...)` in the loop to resolve them.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt (hunk 3 above)
Comment (alwa-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:8]: You can also use some generic form like "Got incorrect return value %d". Or "It incorrectly returned %d". The name of the test is printed above the error during the test run anyway, so adding `__func__` to the error string does not carry any extra information.

**Comment by alwa-nordic** [comment]
> Regarding the note for running clang-format on new sources, Is it a generic remark? or you already found non-formatted coded.
> I ran it on current source files and there were no changes on the code except it's re-ordering the header files inclusions in a way that no any other project use.
> I discarded changes related to header files inclusion re-ordering, please let me know if you think that I should commit the result from clang-format in anyway.

Yes, please format the includes. That was what alerted me in the first place. I want us to use clang-format and/or other formatting tools as much as possible. That way there will be less discussion about formatting in the end.

File: tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt (hunk 1 above)
Comment (alwa-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:13]: Something's not right. I believe the there is a bug in the logging subsystem. When the log it set to go to printf, the empty `BT_DBG("");` should print a newline. So the format-string should not be zero-length. We are hiding a bug by adding this compile option.

Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/bt_id_read_public_addr/CMakeLists.txt:13]: Yes, from functional perspective, I'd expect that `BT_DBG("");` should print a new line, it's not the case.

Also, IMO for that particular case in `id.c`, it's not very helpful and I proposed to delete it, but I got some objections so I went for that workaround to unblock my PR.

**Review by alwa-nordic — DISMISSED** [review]

File: tests/bluetooth/host/id/mocks/hci_core_expects.c [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "mocks/hci_core.h"
+#include "mocks/hci_core_expects.h"
+
+#include <zephyr/bluetooth/buf.h>
+#include <zephyr/kernel.h>
+
+void expect_single_call_bt_unpair(uint8_t id, const bt_addr_le_t *addr)
```
Comment (jori-nordic) [inline tests/bluetooth/host/id/mocks/hci_core_expects.c:13]: this doesn't seem used by anything?

File: tests/bluetooth/host/id/mocks/syscalls/device.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Comment (jori-nordic) [inline tests/bluetooth/host/id/mocks/syscalls/device.h:2]: update copyrights

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: tests/bluetooth/host/id/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+            mocks/scan.c
+            mocks/addr.c
+            mocks/crypto.c
+            mocks/net_buf.c
+            mocks/net_buf_expects.c
+            mocks/hci_core.c
+            mocks/hci_core_expects.c
+
+            ${ZEPHYR_BASE}/subsys/bluetooth/host/id.c
+            ${ZEPHYR_BASE}/subsys/bluetooth/common/addr.c
```
[the remark this answers is not in the document]
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/CMakeLists.txt:15]: I followed modifications done by Alexander to other UTs by adding those files to the sources list of the projects

File: tests/bluetooth/host/id/mocks/hci_core_expects.c (hunk 5 above)
Comment (ahmedmoheb-nordic) [reply tests/bluetooth/host/id/mocks/hci_core_expects.c:13]: It's used by other UTs, but as the same implementation was used for UTs, I copied the whole file at once

Comment (jori-nordic) [reply tests/bluetooth/host/id/mocks/hci_core_expects.c:13]: fair enough

**Review by jori-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
